pkg/scheduler/factory/factory.go

```diff
@@ -1124,21 +1123,11 @@ func (c *configFactory) MakeDefaultErrorFunc(backoff *util.PodBackoff, podQueue
 				if err == nil {
 					if len(pod.Spec.NodeName) == 0 {
 						podQueue.AddUnschedulableIfNotPresent(pod)
-					} else {
-						if c.volumeBinder != nil {
-							// Volume binder only wants to keep unassigned pods
-							c.volumeBinder.DeletePodBindings(pod)
-						}
 					}
 					break
 				}
 				if errors.IsNotFound(err) {
 					klog.Warningf("A pod %v no longer exists", podID)
-
-					if c.volumeBinder != nil {
-						// Volume binder only wants to keep unassigned pods
-						c.volumeBinder.DeletePodBindings(origPod)
```

Comment:
In `deletePodFromSchedulingQueue`:

https://github.com/kubernetes/kubernetes/blob/55d8f709c8b214f72c000076093e08344dc3ebce/pkg/scheduler/factory/factory.go#L580-L603

Reply:
Have revisited `DeletePodBindings` part again. I found `deletePodFromSchedulingQueue` is called only [when a unscheduled pod is deleted](https://github.com/kubernetes/kubernetes/blob/55d8f709c8b214f72c000076093e08344dc3ebce/pkg/scheduler/factory/factory.go#L314-L338).

After this PR, we do not need to clear pod binding cache on error because we recreate at beginning of scheduler loop, but we had never cleared cache on success.

IMO, pod binding cache at least should be cleared when pod is not needed to be scheduled (e.g. assigned a node on success or deleted from cluster).

Don't want to make this PR too complex, I created a issue to track this: https://github.com/kubernetes/kubernetes/issues/72763

Reply:
Sorry, I misunderstood [this part code](https://github.com/kubernetes/kubernetes/blob/55d8f709c8b214f72c000076093e08344dc3ebce/pkg/scheduler/factory/factory.go#L314-L338), `deletePodFromSchedulingQueue` will be called on success, because with FilteringResourceEventHandler, object stops passing the filter (assigned a pod or not here) after an update is considered a delete. So all successful pods will go through `deletePodFromSchedulingQueue`.

Talked with @msau42 offline, we'll add integration tests for verify no memory leaks on pod binding cache.